docs/spanner.md

```diff
@@ -0,0 +1,189 @@
+
+# Annotate the kubectl serviceAccountName with the GCP Service Account
+kubectl annotate serviceaccount ${KSA_NAME} \
+    --namespace=${NAMESPACE} \
+    iam.gke.io/gcp-service-account=${GSA_NAME}
+
+# Tell gcloud that the kubectl account maps to the GCP one
+gcloud iam service-accounts add-iam-policy-binding \
+    "${GSA_NAME}" --role roles/iam.workloadIdentityUser \
+    --member "serviceAccount:${PROJECT_ID}.svc.id.goog[${NAMESPACE}/${KSA_NAME}]"
+```
+
+### 5. Configure the Kubernetes manifest to use Spanner for the `cartservice`.
+
+```sh
+cp ./release/kubernetes-manifests.yaml ./release/updated-manifests.yaml
+sed -i "s/name: REDIS_ADDR/name: SPANNER_PROJECT/g" ./release/updated-manifests.yaml
+sed -i "s/value: \"redis-cart:6379\"/value: \"${PROJECT_ID}\"/g" ./release/updated-manifests.yaml
+sed -i "s/cartservice:v0.4.0/cartservice:v0.4.0-spanner/g" ./release/updated-manifests.yaml
+```
+
+#### *Note on Spanner connection environmental variables*
+
+You can add also manually edit the `updated-manifests.yaml`
+`cartservice` section, and add environmental variables to specify a Spanner instance, database, or an explicit connection string.
+The following environmental variables will be used by the `cartservice`, if present:
+
+- `SPANNER_INSTANCE`: defaults to `onlineboutique`, unless specified.
+- `SPANNER_DATABASE`: defaults to `carts`, unless specified.
+- `SPANNER_CONNECTION_STRING`: defaults to `projects/${SPANNER_PROJECT}/instances/${SPANNER_INSTANCE}/databases/${SPANNER_DATABASE}`. If this variable is defined explicitly, all other environmental variables will be ignored.
+
+### 6. Apply the updated manifest
+
+```sh
+kubectl apply -f ./release/updated-manifests.yaml
+```
+
+#### Optional: delete the unused redis-cart service
+
+```sh
+kubectl delete deployment redis-cart
+```
+
+### 7. Wait for the Pods to be ready
+
+```sh
+kubectl get pods
+```
+
+After a few minutes, you should see something like this:
+
+```
+NAME                                     READY   STATUS    RESTARTS   AGE
+adservice-76bdd69666-ckc5j               1/1     Running   0          2m58s
+cartservice-66d497c6b7-dp5jr             1/1     Running   0          2m59s
+checkoutservice-666c784bd6-4jd22         1/1     Running   0          3m1s
+currencyservice-5d5d496984-4jmd7         1/1     Running   0          2m59s
+emailservice-667457d9d6-75jcq            1/1     Running   0          3m2s
+frontend-6b8d69b9fb-wjqdg                1/1     Running   0          3m1s
+loadgenerator-665b5cd444-gwqdq           1/1     Running   0          3m
+paymentservice-68596d6dd6-bf6bv          1/1     Running   0          3m
+productcatalogservice-557d474574-888kr   1/1     Running   0          3m
+recommendationservice-69c56b74d4-7z8r5   1/1     Running   0          3m1s
+shippingservice-6ccc89f8fd-v686r         1/1     Running   0          2m58s
+```
+
+### 8. Access the web frontend in a browser
+
+```sh
+kubectl get service frontend-external | awk '{print $4}'
+```
```

Comment:
**Issue:**

When I deployed Online Boutique with Cloud Spanner by following the instructions here and using a Cloud-Spanner supporting image, I got the following errors:

```
rpc error: code = FailedPrecondition desc = Can't access cart storage at projects/nimjay-ob-spanner-pr/instances/onlineboutique/databases/carts. System.ArgumentException: Format of the initialization string does not conform to specification starting at index 0.
   at System.Data.Common.DbConnectionOptions.GetKeyValuePair(String, Int32, StringBuilder, Boolean, String& , String& )
   at System.Data.Common.DbConnectionOptions.ParseInternal(Dictionary`2, String, Boolean, Dictionary`2, Boolean)
   at System.Data.Common.DbConnectionOptions..ctor(String, Dictionary`2, Boolean)
   at System.Data.Common.DbConnectionStringBuilder.set_ConnectionString(String value)
   at cartservice.cartstore.SpannerCartStore.GetCartAsync(String) in /app/cartstore/SpannerCartStore.cs:line 139
could not retrieve cart
main.(*frontendServer).viewCartHandler
	/src/handlers.go:255
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2109
github.com/gorilla/mux.(*Router).ServeHTTP
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210
main.(*logHandler).ServeHTTP
	/src/middleware.go:82
main.ensureSessionID.func1
	/src/middleware.go:109
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2109
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2947
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1991
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1594
```

and 

```
rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.108.9.216:7070: connect: connection refused"
could not retrieve cart
main.(*frontendServer).viewCartHandler
	/src/handlers.go:255
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2109
github.com/gorilla/mux.(*Router).ServeHTTP
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210
main.(*logHandler).ServeHTTP
	/src/middleware.go:82
main.ensureSessionID.func1
	/src/middleware.go:109
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2109
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2947
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1991
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1594
```

It's been about 20 minutes since I created my Cloud Spanner instance.
Do you have any idea what might be causing this?

Reply:
For clarity, here are some screenshots:
![Screen Shot 2022-10-05 at 11 06 17 AM](https://user-images.githubusercontent.com/10292865/194095035-cf57eea5-47aa-4c09-8a94-8a107b094753.png)
![Screen Shot 2022-10-05 at 11 00 42 AM](https://user-images.githubusercontent.com/10292865/194095057-20bf118c-61dc-4710-9313-82104a73e5d0.png)

Reply:
I think I figure it out: I think it's because I didn't specific `--project=${PROJECT_ID}` in step 4. 😅 
Let me retry. :)

Reply:
Update: Still getting the same errors.

Reply:
Thanks @NimJay for this thorough review! I know the `spanner.md` walkthrough is the most time-consuming, but also most effective way to test this whole thing out.

Guessing by the subsequent approval, it looks like you got it to work, but I'm happy to help if you hit any other snags.

Reply:
I'm getting the error with the image at [gcr.io/google-samples/microservices-demo/cartservice:v0.4.0-spanner](https://gcr.io/google-samples/microservices-demo/cartservice:v0.4.0-spanner), but not with the one in my private repo. I'll try rebuilding mine and see if that can reproduce the issue.

Reply:
(Please don't feel obligated to respond outside your work hours! :))

Just leaving an update:
I added "Data Source=" to the `connectionString` (based on [this documentation](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.Data/latest/Google.Cloud.Spanner.Data.SpannerConnection#Google_Cloud_Spanner_Data_SpannerConnection__ctor_System_String_Grpc_Core_ChannelCredentials_)). See commit https://github.com/GoogleCloudPlatform/microservices-demo/pull/1109/commits/8b78843739108a110201588aa3256791fb8ab922.

Now I see the following error:
```
rpc error: code = FailedPrecondition desc = Can't access cart storage at Data Source=projects/nimjay-ob-spanner-pr/instances/onlineboutique/databases/carts. Google.Cloud.Spanner.Data.SpannerException: Internal error.
 ---> Grpc.Core.RpcException: Status(StatusCode="Internal", Detail="Error starting gRPC call. TokenResponseException: Error:"Server response does not contain a JSON object. Status code is: BadRequest", Description:"", Uri:""", DebugException="Google.Apis.Auth.OAuth2.Responses.TokenResponseException: Error:"Server response does not contain a JSON object. Status code is: BadRequest", Description:"", Uri:""
   at Google.Apis.Auth.OAuth2.Responses.TokenResponse.FromHttpResponseAsync(HttpResponseMessage, IClock, ILogger)
   at Google.Apis.Auth.OAuth2.ComputeCredential.RequestAccessTokenAsync(CancellationToken)
   at Google.Apis.Auth.OAuth2.TokenRefreshManager.RefreshTokenAsync()
   at Google.Apis.Auth.OAuth2.TokenRefreshManager.GetAccessTokenForRequestAsync(CancellationToken)
   at Google.Apis.Auth.OAuth2.ServiceCredential.GetAccessTokenWithHeadersForRequestAsync(String , CancellationToken )
   at Grpc.Auth.GoogleAuthInterceptors.<>c__DisplayClass3_0.<<FromCredential>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Grpc.Net.Client.Internal.GrpcProtocolHelpers.ReadCredentialMetadata(DefaultCallCredentialsConfigurator, GrpcChannel, HttpRequestMessage, IMethod, CallCredentials)
   at Grpc.Net.Client.Internal.GrpcCall`2.ReadCredentials(HttpRequestMessage)
   at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage, Nullable`1)")
   at Google.Api.Gax.Grpc.Gcp.GcpCallInvoker.<>c__DisplayClass24_0`2.<<AsyncUnaryCall>g__PostProcessPropagateResult|3>d.MoveNext()
--- End of stack trace from previous location ---
   at Google.Api.Gax.Grpc.ApiCallRetryExtensions.<>c__DisplayClass0_0`2.<<WithRetry>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Google.Cloud.Spanner.V1.SessionPool.TargetedSessionPool.CreatePooledSessionsAsync(CancellationToken)
   at Google.Cloud.Spanner.V1.SessionPool.TargetedSessionPool.<>c__DisplayClass43_0.<<GetNursePoolBackToHealthTask>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Google.Cloud.Spanner.V1.SessionPool.TargetedSessionPool.<GetNursePoolBackToHealthTask>g__WithCancellationTokenAsync|43_1[TTaskResult](Task`1, CancellationToken)
   at Google.Cloud.Spanner.V1.SessionPool.TargetedSessionPool.<>c__DisplayClass31_0.<<GetSessionAcquisitionTask>g__NurseAndRetryAsync|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Google.Cloud.Spanner.V1.SessionPool.TargetedSessionPool.AcquireSessionImplAsync(TransactionOptions, CancellationToken)
   at Google.Cloud.Spanner.V1.SessionPool.TargetedSessionPool.AcquireSessionAsync(TransactionOptions, CancellationToken)
   at Google.Cloud.Spanner.Data.EphemeralTransaction.<>c__DisplayClass7_0.<<ExecuteReadOrQueryAsync>g__Impl|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Google.Cloud.Spanner.Data.ExecuteHelper.WithErrorTranslationAndProfiling[T](Func`1, String, Logger)
   --- End of inner exception stack trace ---
   at Google.Cloud.Spanner.Data.ExecuteHelper.WithErrorTranslationAndProfiling[T](Func`1, String, Logger)
   at Google.Cloud.Spanner.Data.SpannerCommand.ExecutableCommand.ExecuteReaderAsync(CommandBehavior, TimestampBound, CancellationToken)
   at Google.Cloud.Spanner.Data.SpannerCommand.ExecutableCommand.ExecuteDbDataReaderAsync(CommandBehavior, TimestampBound, CancellationToken)
   at Google.Cloud.Spanner.Data.SpannerCommand.ExecuteReaderAsync(CommandBehavior, CancellationToken)
   at cartservice.cartstore.SpannerCartStore.GetCartAsync(String) in /app/cartstore/SpannerCartStore.cs:line 116
could not retrieve cart
main.(*frontendServer).homeHandler
	/src/handlers.go:69
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2109
github.com/gorilla/mux.(*Router).ServeHTTP
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210
main.(*logHandler).ServeHTTP
	/src/middleware.go:82
main.ensureSessionID.func1
	/src/middleware.go:109
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:2109
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2947
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1991
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1594
```

Reply:
I reran the commands from step 4 (related to service accounts). The above error disappeared. Online Boutique with Spanner is now working for me! 👍